service/frontend/task_reachability.go

```diff
@@ -0,0 +1,299 @@
+		compatibilityResponse, err := request.versionSetFetcher.fetchTaskQueueVersions(ctx, request.namespace, taskQueue)
+		if err != nil {
+			return nil, err
+		}
+		return wh.getTaskQueueReachability(ctx, taskQueueReachabilityRequest{
+			buildId:          request.buildId,
+			namespace:        request.namespace,
+			taskQueue:        taskQueue,
+			versionSets:      compatibilityResponse.GetResponse().GetMajorVersionSets(),
+			reachabilityType: request.reachabilityType,
+		})
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(taskQueuesToSkip) > 0 {
+		skippedTasksReachability := make([]*taskqueuepb.TaskQueueReachability, len(taskQueuesToSkip))
+		for i, taskQueue := range taskQueuesToSkip {
+			skippedTasksReachability[i] = &taskqueuepb.TaskQueueReachability{TaskQueue: taskQueue, Reachability: []enumspb.TaskReachability{enumspb.TASK_REACHABILITY_UNSPECIFIED}}
+		}
+		taskQueueReachability = append(taskQueueReachability, skippedTasksReachability...)
+	}
+	return &taskqueuepb.BuildIdReachability{BuildId: request.buildId, TaskQueueReachability: taskQueueReachability}, nil
+}
+
+type taskQueueReachabilityRequest struct {
+	buildId          string
+	taskQueue        string
+	namespace        *namespace.Namespace
+	versionSets      []*taskqueuepb.CompatibleVersionSet
+	reachabilityType enumspb.TaskReachability
+}
+
+// Get the reachability of a single build ID in a single task queue scope.
+func (wh *WorkflowHandler) getTaskQueueReachability(ctx context.Context, request taskQueueReachabilityRequest) (*taskqueuepb.TaskQueueReachability, error) {
+	taskQueueReachability := taskqueuepb.TaskQueueReachability{TaskQueue: request.taskQueue, Reachability: []enumspb.TaskReachability{}}
+
+	var isDefaultInQueue bool
+	var buildIdsFilter string
+
+	if request.buildId == "" {
+		// Query for the unversioned worker
+		isDefaultInQueue = len(request.versionSets) == 0
+		// Query workflows that have completed tasks marked with a sentinel "unversioned" search attribute.
+		buildIdsFilter = fmt.Sprintf("BuildIds = %q", common.UnversionedSearchAttribute)
+	} else {
+		// Query for a versioned worker
+		setIdx := -1
+		buildIdIdx := -1
+		if len(request.versionSets) > 0 {
+			for sidx, set := range request.versionSets {
+				for bidx, id := range set.BuildIds {
+					if request.buildId == id {
+						setIdx = sidx
+						buildIdIdx = bidx
+						break
+					}
+				}
+			}
+		}
+
+		if setIdx == -1 {
+			// build id not in set - unreachable
+			return &taskQueueReachability, nil
+		}
+		set := request.versionSets[setIdx]
+		isDefaultInSet := buildIdIdx == len(set.BuildIds)-1
+
+		if !isDefaultInSet {
+			// unreachable
+			return &taskQueueReachability, nil
+		}
+
+		isDefaultInQueue = setIdx == len(request.versionSets)-1
+		escapedBuildIds := make([]string, len(set.GetBuildIds()))
+		for i, buildId := range set.GetBuildIds() {
+			escapedBuildIds[i] = fmt.Sprintf("%q", common.VersionedBuildIdSearchAttribute(buildId))
```

Comment:
use quoting and escaping from the sql parser, not from go

Reply:
I thought the Go one was safe, I guess not...
Do you have an example I for reference?

Reply:
we use `github.com/xwb1989/sqlparser` for parsing queries. disappointingly there's no simple "quote" function but I figured out you can do it like this: https://go.dev/play/p/MzUip_xtkiM